transaction/DeadLockDetector.cpp

```diff
@@ -0,0 +1,186 @@
+ **/
+
+#include "transaction/DeadLockDetector.hpp"
+
+#include <algorithm>
+#include <atomic>
+#include <chrono>
+#include <memory>
+#include <utility>
+#include <vector>
+
+#include "transaction/CycleDetector.hpp"
+#include "transaction/DirectedGraph.hpp"
+#include "transaction/LockTable.hpp"
+#include "transaction/Transaction.hpp"
+
+namespace quickstep {
+namespace transaction {
+
+constexpr std::int64_t DeadLockDetector::kSleepDuration;
+
+DeadLockDetector::DeadLockDetector(LockTable *lock_table,
+                                   std::atomic<DeadLockDetectorStatus> *status,
+                                   std::vector<DirectedGraph::node_id> *victims)
+    : wait_for_graph_(nullptr),
+      tid_node_mapping_(std::make_unique<transaction_id_node_map>()),
+      lock_table_(lock_table),
+      status_(*status),
+      victims_(*victims) {
+}
+
+void DeadLockDetector::run() {
+  while (true) {
+    if (status_.load() == DeadLockDetectorStatus::kQuit) {
+      // DeadLockDetector should stop.
+      return;
+    }
+    while (status_.load() == DeadLockDetectorStatus::kDone) {
+      // LockTable has not process the previous batch yet.
+    }
+
+    // TODO(Hakan): Implement logging mechanism for deadlock detection
+    //              start and end times.
+    std::vector<DirectedGraph::node_id> victim_new_batch = getAllVictims();
+
+    // Swap new batch with old batch to make LockTable to see new victims.
+    std::swap(victim_new_batch, victims_);
+
+    // Signal LockTable that new batch is ready.
+    status_.store(DeadLockDetectorStatus::kDone);
+
+    // DeadLockDetector should run once in a predefined interval.
+    std::this_thread::sleep_for(
+        std::chrono::seconds(kSleepDuration));
+  }
+}
+
+void DeadLockDetector::addPendingInfo(transaction_id pending,
+                                      transaction_id owner) {
+  DirectedGraph::node_id pending_node_id = getNodeId(pending);
+  DirectedGraph::node_id owner_node_id = getNodeId(owner);
+
+  // TODO(Hakan): Check first whether link is already created. Use checked
+  //              version for adding an edge.
+  wait_for_graph_->addEdgeUnchecked(pending_node_id, owner_node_id);
+}
+
+void DeadLockDetector::deletePendingInfo(transaction_id pending,
+                                         transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+bool DeadLockDetector::isDependent(transaction_id pending, transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+std::vector<transaction_id>
+DeadLockDetector::getAllDependents(transaction_id owner) {
+  FATAL_ERROR("Not implemented");
+}
+
+std::vector<transaction_id>
+DeadLockDetector::getAllDependees(transaction_id pending) {
+  DirectedGraph::node_id pending_node_id = getNodeId(pending);
+  std::vector<DirectedGraph::node_id> nodes
+      = wait_for_graph_->getAdjacentNodes(pending_node_id);
+  std::vector<transaction_id> transactions;
+  transactions.reserve(nodes.size());
```

Comment:
I don't understand why you first reserve some places, but then append more elements in the end.